pkg/audit/verifier.go

```diff
@@ -225,32 +226,40 @@ func (verifier *Verifier) Verify(ctx context.Context, stripe *Stripe, skip map[s
 }
 
 // DownloadShares downloads shares from the nodes where remote pieces are located
-func (verifier *Verifier) DownloadShares(ctx context.Context, limits []*pb.AddressedOrderLimit, stripeIndex int64, shareSize int32) (shares map[int]Share, nodes map[int]storj.NodeID, err error) {
+func (verifier *Verifier) DownloadShares(ctx context.Context, limits []*pb.AddressedOrderLimit, stripeIndex int64, shareSize int32) (shares map[int]Share, err error) {
 	defer mon.Task()(&ctx)(&err)
 
 	shares = make(map[int]Share, len(limits))
-	nodes = make(map[int]storj.NodeID, len(limits))
+	ch := make(chan *Share, len(limits))
 
 	for i, limit := range limits {
 		if limit == nil {
+			ch <- nil
 			continue
 		}
 
-		// TODO(kaloyan): execute in goroutine for better performance
-		share, err := verifier.GetShare(ctx, limit, stripeIndex, shareSize, i)
-		if err != nil {
-			share = Share{
-				Error:    err,
-				PieceNum: i,
-				Data:     nil,
+		go func(i int, limit *pb.AddressedOrderLimit) {
+			share, err := verifier.GetShare(ctx, limit, stripeIndex, shareSize, i)
+			if err != nil {
+				share = Share{
+					Error:    err,
+					PieceNum: i,
+					NodeID:   limit.GetLimit().StorageNodeId,
+					Data:     nil,
+				}
 			}
-		}
+			ch <- &share
+		}(i, limit)
+	}
 
-		shares[share.PieceNum] = share
-		nodes[share.PieceNum] = limit.GetLimit().StorageNodeId
+	for range limits {
+		share := <-ch
+		if share != nil {
```

Comment:
Can a share be `nil` here? or only if there is a bug in `verifier.GetShare`?

Reply:
Yes. Look at line 237. `limits` is ordered by piece number. So it can have `nil` values for those piece numbers that no storage node is storing a piece.